numpy/core/src/multiarray/compiled_base.c

```diff
@@ -302,193 +302,152 @@ arr_digitize(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwds)
-                rem_indx /= inshape[i];
-            }
-            indx += rem_indx * instrides[0];
-            /* fprintf(stderr, "mindx = %d, indx=%d\n", mindx, indx); */
-            /* Copy value element over to input array */
-            memcpy(input_data+indx,vptr,delsize);
-            if (objarray) {
-                Py_INCREF(*((PyObject **)vptr));
-            }
-            vptr += delsize;
-            copied += 1;
-            /* If we move past value data.  Reset */
-            if (copied >= numvals) {
-                vptr = avals_data;
-                copied = 0;
-            }
-        }
-        mptr += melsize;
-    }
-}
-
-/*
  * Returns input array with values inserted sequentially into places
  * indicated by the mask
  */
 NPY_NO_EXPORT PyObject *
 arr_insert(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
 {
-    PyObject *mask = NULL, *vals = NULL;
-    PyArrayObject *ainput = NULL, *amask = NULL, *avals = NULL, *tmp = NULL;
-    int numvals, totmask, sameshape;
-    char *input_data, *mptr, *vptr, *zero = NULL;
-    int melsize, delsize, nd, objarray, k;
-    npy_intp *instrides, *inshape;
+    int copied;
+    char *src, *dest;
+    npy_bool *mask_data;
+    PyArray_Descr *dtype;
+    PyObject *array0, *mask0, *values0;
+    PyArrayObject *array, *mask, *values;
+    npy_intp i, j, chunk, ni, max_item, nv;
 
     static char *kwlist[] = {"input", "mask", "vals", NULL};
 
-    if (!PyArg_ParseTupleAndKeywords(args, kwdict, "O&OO", kwlist,
-                PyArray_Converter, &ainput,
-                &mask, &vals)) {
-        goto fail;
+    if (!PyArg_ParseTupleAndKeywords(args, kwdict, "O!OO:place", kwlist,
+                &PyArray_Type, &array0, &mask0, &values0)) {
+        return NULL;
     }
 
-    amask = (PyArrayObject *)PyArray_FROM_OF(mask, NPY_ARRAY_CARRAY);
-    if (amask == NULL) {
-        goto fail;
-    }
-    /* Cast an object array */
-    if (PyArray_DESCR(amask)->type_num == NPY_OBJECT) {
-        tmp = (PyArrayObject *)PyArray_Cast(amask, NPY_INTP);
-        if (tmp == NULL) {
-            goto fail;
-        }
-        Py_DECREF(amask);
-        amask = tmp;
-    }
+    copied = 0;
+    mask = NULL;
+    values = NULL;
+    array = (PyArrayObject *)array0;
 
-    sameshape = 1;
-    if (PyArray_NDIM(amask) == PyArray_NDIM(ainput)) {
-        for (k = 0; k < PyArray_NDIM(amask); k++) {
-            if (PyArray_DIMS(amask)[k] != PyArray_DIMS(ainput)[k]) {
-                sameshape = 0;
-            }
-        }
+    if (!PyArray_Check(array)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "place: first argument must "
+                        "be an array");
+        return NULL;
     }
-    else {
-        /* Test to see if amask is 1d */
-        if (PyArray_NDIM(amask) != 1) {
-            sameshape = 0;
-        }
-        else if ((PyArray_SIZE(ainput)) != PyArray_SIZE(amask)) {
-            sameshape = 0;
+
+    if (!PyArray_ISCONTIGUOUS(array)) {
```

Comment:
This block can be simplified a lot, I think...

 1. You don't need to explicitly handle dtype ref counting: just pass `NULL` to `PyArray_FromArray` and it will do it for you.
 2. If inside this `if` block, `obj != array` will always be true, as `array` isn't contiguous but `obj` is, so that check is unnecessary.
 3. You do need to check that `obj` is not `NULL`!

I think a better approach would be to unconditionally call `PyArray_FromArray` and get rid of the `copied` flag. If the array already fulfills the conditions, it will be INCREF'ed, if not a new array will be created. Either way, you need to DECREF it before returning.  So something along these lines:

```C
    PyArrayObject *array = NULL;

...

    array = (PyArrayObject *)PyArray_FromArray((PyArrayObject *)array0, NULL,
                                NPY_ARRAY_CARRAY | NPY_ARRAY_UPDATEIFCOPY);
    if (array == NULL) {
        goto fail;
    }

...

    Py_DECREF(array);
    Py_RETURN_NONE;
fail:
    Py_XDECREF(array);

...
```